friendships for two rows only one invitation allowed

gonjavi opened the pull request:
**Milestone 6: Friendships v2**

In this milestone we:
- Added only one friend's invitation is allowed.

- Moved the friendship#index page to users#index to render friends, pending requests, friend requests, and suggested friends.

- Removed some logic from posts#index view, adding the method friends_posts_and_own_posts in the user model.

- Removed some logic from friendships#index which is now users#index, adding methods show_pending_friends and show_friend_requests in the user models.

- Modified the method confirm_friend in the user model.

- Added the methods confirm_friendship(user), show_friendship_requests(user), and show_friendships(pending_user) to be used in the users#index view.

- Removed test "it renders the index template" (RSpec) in friendships_controller_spec and moved it to users_controller_spec.

maelfosso commented:
https://github.com/gonjavi/RailsFacebookApp/blob/milestone6/app/models/user.rb#L37-L41

gonjavi commented:
Ok, We are going to work on it

gonjavi commented:
The changes were done taking into account the review, and I have done point 4 in a different way, some things were done as close as possible, and other ones slightly different as I could have the result.

Torres-ssf commented on app/models/user.rb:
```diff
@@ -24,18 +24,44 @@ def sug_users
+  def confirm_friendship(user)
+    confirm_f_array = inverse_friendships.map { |f| f if f.user == user and f.friend == self }
+    confirm_f_array.compact
   end
```
**OPTIONAL**
We can use the Active Record interface to query information instead of ruby methods. Active Record methods are optimized to query information and do that in a faster way then ruby methods. Check out this [link](https://guides.rubyonrails.org/active_record_querying.html) to learn more about it

gonjavi replied:
Thanks,
Good to know that, in this project, I have used ruby methods from the previous milestone taking  into account the suggested article: https://smartfunnycool.com/friendships-in-activerecord/
In this milestone, I learned how to create my own methods to achieve the desired results in this project, in future projects I will use Active Record to query information in a faster way.

maelfosso commented on app/controllers/posts_controller.rb:
```diff
@@ -5,12 +5,10 @@ class PostsController < ApplicationController
   def index
-    @posts = Post.all.order('created_at DESC').all
-    @comments = Comment.all.order('created_at ASC').all
+    @posts = Post.all
```
Why do you need **all** the posts?

gonjavi replied:
thanks
according to point 1. Move the friendship confirmation into the Friendship model, if I move it, it does not work

gonjavi commented:
Changes added in the previous commit:

- Changed @posts by current_user.friends_posts_and_own_posts in posts index view in the line 180 (166 now)of the js part, and removed @post = Post.all in posts controller index method.

- Removed the **if current_user**, and **the else - end** block from posts index view, and created a new root page posts#home with that content. 

- In the posts_controller added the method home, and added rule in before action so that the method index can be accessed only by the authenticated user.

- If the user is not authenticated is sent to the root page, which is posts home view.

- Added after_sign_in method in application controller to redirect to posts index as soon as the user signs in.

samgaco commented on app/models/friendship.rb:
```diff
@@ -4,4 +4,10 @@ class Friendship < ApplicationRecord
+  def confirm_friend(user)
+    friendship = inverse_friendships.find { |f| f.user == user and f.friend == self }
+    friendship.confirmed = true
+    friendship.save
```
- [ ] When a friendship is accepted, a new row with the inverse relation of the friendship should be created.

gonjavi replied:
Perfect, thanks

gonjavi commented:
Changes added in the previous commit:

- When a friendship is accepted, it creates a new row with an inverse relation of friendship. I have Modified the method confirm_friend(user) in the corresponding model to achieve this result.

- Modified destroy method in friendship controller to delete the friendship and its inverse friendship.
